Cubical/Data/SumFin/Properties.agda

```diff
@@ -87,7 +92,179 @@ SumFin∥∥≃ (suc n) =
+trueCount {n = 0} _ = 0
+trueCount {n = suc n} f = Bool→ℕ (f (inl tt)) + (trueCount (f ∘ inr))
+
+SumFinDec⊎≃ : (n : ℕ)(t : Bool) → (Bool→Type t ⊎ Fin n) ≃ (Fin (Bool→ℕ t + n))
+SumFinDec⊎≃ _ true = idEquiv _
+SumFinDec⊎≃ _ false = ⊎-swap-≃ ⋆ ⊎-⊥-≃
+
+SumFinSub≃ : (n : ℕ)(f : Fin n → Bool) → Σ _ (Bool→Type ∘ f) ≃ Fin (trueCount f)
+SumFinSub≃ 0 _ = ΣEmpty _
+SumFinSub≃ (suc n) f =
+    Σ⊎≃
+  ⋆ ⊎-equiv (ΣUnit (Bool→Type ∘ f ∘ inl)) (SumFinSub≃ n (f ∘ inr))
+  ⋆ SumFinDec⊎≃ _ (f (inl tt))
+
+-- decidable quantifier
+
+trueForSome : (n : ℕ)(f : Fin n → Bool) → Bool
+trueForSome 0 _ = false
+trueForSome (suc n) f = f (inl tt) or trueForSome n (f ∘ inr)
+
+trueForAll : (n : ℕ)(f : Fin n → Bool) → Bool
+trueForAll 0 _ = true
+trueForAll (suc n) f = f (inl tt) and trueForAll n (f ∘ inr)
+
+SumFin∃→ : (n : ℕ)(f : Fin n → Bool) → Σ _ (Bool→Type ∘ f) → Bool→Type (trueForSome n f)
+SumFin∃→ 0 _ = ΣEmpty _ .fst
+SumFin∃→ (suc n) f =
+    Bool→Prop⊎' _ _
+  ∘ map-⊎ (ΣUnit (Bool→Type ∘ f ∘ inl) .fst) (SumFin∃→ n (f ∘ inr))
+  ∘ Σ⊎≃ .fst
+
+SumFin∃← : (n : ℕ)(f : Fin n → Bool) → Bool→Type (trueForSome n f) → Σ _ (Bool→Type ∘ f)
+SumFin∃← 0 _ = ⊥.rec
+SumFin∃← (suc n) f =
+    invEq Σ⊎≃
+  ∘ map-⊎ (invEq (ΣUnit (Bool→Type ∘ f ∘ inl))) (SumFin∃← n (f ∘ inr))
+  ∘ Bool→Prop⊎ _ _
+
+SumFin∃≃ : (n : ℕ)(f : Fin n → Bool) → ∥ Σ (Fin n) (Bool→Type ∘ f) ∥ ≃ Bool→Type (trueForSome n f)
+SumFin∃≃ n f =
+  propBiimpl→Equiv isPropPropTrunc isPropBool→Prop
+    (Prop.rec isPropBool→Prop (SumFin∃→ n f))
+    (∣_∣ ∘ SumFin∃← n f)
+
+SumFin∀≃ : (n : ℕ)(f : Fin n → Bool) → ((x : Fin n) → Bool→Type (f x)) ≃ Bool→Type (trueForAll n f)
+SumFin∀≃ 0 _ = isContr→≃Unit (isContrΠ⊥)
+SumFin∀≃ (suc n) f =
+    Π⊎≃
+  ⋆ Σ-cong-equiv (ΠUnit (Bool→Type ∘ f ∘ inl)) (λ _ → SumFin∀≃ n (f ∘ inr))
+  ⋆ Bool→Prop≃ _ _
+
+-- internal equality
+
+SumFin≡ : (n : ℕ) → (a b : Fin n) → Bool
+SumFin≡ 0 _ _ = true
+SumFin≡ (suc n) (inl tt) (inl tt) = true
+SumFin≡ (suc n) (inl tt) (inr y) = false
+SumFin≡ (suc n) (inr x) (inl tt) = false
+SumFin≡ (suc n) (inr x) (inr y) = SumFin≡ n x y
+
+isSetSumFin : (n : ℕ)→ isSet (Fin n)
+isSetSumFin 0 = isProp→isSet isProp⊥
+isSetSumFin (suc n) = isSet⊎ (isProp→isSet isPropUnit) (isSetSumFin n)
+
+SumFin≡≃ : (n : ℕ) → (a b : Fin n) → (a ≡ b) ≃ Bool→Type (SumFin≡ n a b)
+SumFin≡≃ 0 _ _ = isContr→≃Unit (isProp→isContrPath isProp⊥ _ _)
+SumFin≡≃ (suc n) (inl tt) (inl tt) = isContr→≃Unit (inhProp→isContr refl (isSetSumFin _ _ _))
+SumFin≡≃ (suc n) (inl tt) (inr y) = invEquiv (⊎Path.Cover≃Path _ _) ⋆ uninhabEquiv ⊥.rec* ⊥.rec
+SumFin≡≃ (suc n) (inr x) (inl tt) = invEquiv (⊎Path.Cover≃Path _ _) ⋆ uninhabEquiv ⊥.rec* ⊥.rec
+SumFin≡≃ (suc n) (inr x) (inr y) = invEquiv (_ , isEmbedding-inr x y) ⋆ SumFin≡≃ n x y
+
+-- propositional truncation of Fin
+
+-- decidability of Fin
+
+DecFin : (n : ℕ) → Dec (Fin n)
+DecFin 0 = no (idfun _)
+DecFin (suc n) = yes fzero
+
+-- propositional truncation of Fin
+
+Dec∥Fin∥ : (n : ℕ) → Dec ∥ Fin n ∥
+Dec∥Fin∥ n = Dec∥∥ (DecFin n)
```

Comment:
Some repetition here, I think. You already have `SumFin∥∥Dec` above.

Reply:
Yeah, but they are different. They use different formu;ation of decidable propositions.

Reply:
The `SumFin∥∥Dec` in this file? As far as I can tell they have slightly different definitions but exactly the same type.

Reply:
Oh you're right...

Reply:
I delete `SumFin∥∥Dec` since it is used nowhere in the library.